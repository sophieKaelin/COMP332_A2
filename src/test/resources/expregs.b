//
// This file is part of COMP332 Assignment 2/3 2019.
//
// weBCPL, a retro BCPL to WebAssembly compiler.
//
// Â© 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// BCPL expression examples, to illustrate the precedence and
// associativity rules of the language.
//

GLOBAL {
    start : 1
}

LET x, y, i, j = ?, ?, 3, 5

AND v = VEC 49

AND start() = VALOF {
    LET a = ! @ x
    AND b = ! v ! i ! j
    AND c = @ v ! i ! j
    AND d = x << 1+y >> 1
    AND e = ~ x ! y
    AND f = ~ x = y
    AND g = NOT x = y
    AND h = - 10 * a / b + c - -10 + d * e
    AND k = FALSE XOR f & g | a & h | c EQV i + j
    AND l = v!i + v!j - a * b

    RESULTIS a -> b, c -> d, e
}