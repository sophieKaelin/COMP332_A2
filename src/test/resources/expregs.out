Program(
    Vector(
        GlobalDecl(Vector(GlobalEntry(IdnDef("start"), Some(IntExp(1))))),
        LetDecl(
            Vector(
                LetVarClause(
                    Vector(IdnDef("x"), IdnDef("y"), IdnDef("i"), IdnDef("j")),
                    Vector(UndefExp(), UndefExp(), IntExp(3), IntExp(5))),
                LetVecClause(IdnDef("v"), IntExp(49)),
                LetFnClause(
                    IdnDef("start"),
                    Vector(),
                    ValofExp(
                        Block(
                            Vector(
                                LetDecl(
                                    Vector(
                                        LetVarClause(
                                            Vector(IdnDef("a")),
                                            Vector(UnaryPlingExp(AddrOfExp(IdnExp(IdnUse("x")))))),
                                        LetVarClause(
                                            Vector(IdnDef("b")),
                                            Vector(
                                                UnaryPlingExp(
                                                    BinaryPlingExp(
                                                        BinaryPlingExp(
                                                            IdnExp(IdnUse("v")),
                                                            IdnExp(IdnUse("i"))),
                                                        IdnExp(IdnUse("j")))))),
                                        LetVarClause(
                                            Vector(IdnDef("c")),
                                            Vector(
                                                AddrOfExp(
                                                    BinaryPlingExp(
                                                        BinaryPlingExp(
                                                            IdnExp(IdnUse("v")),
                                                            IdnExp(IdnUse("i"))),
                                                        IdnExp(IdnUse("j")))))),
                                        LetVarClause(
                                            Vector(IdnDef("d")),
                                            Vector(
                                                ShiftRightExp(
                                                    ShiftLeftExp(
                                                        IdnExp(IdnUse("x")),
                                                        PlusExp(IntExp(1), IdnExp(IdnUse("y")))),
                                                    IntExp(1)))),
                                        LetVarClause(
                                            Vector(IdnDef("e")),
                                            Vector(
                                                NotExp(
                                                    BinaryPlingExp(
                                                        IdnExp(IdnUse("x")),
                                                        IdnExp(IdnUse("y")))))),
                                        LetVarClause(
                                            Vector(IdnDef("f")),
                                            Vector(
                                                NotExp(
                                                    EqualExp(
                                                        IdnExp(IdnUse("x")),
                                                        IdnExp(IdnUse("y")))))),
                                        LetVarClause(
                                            Vector(IdnDef("g")),
                                            Vector(
                                                NotExp(
                                                    EqualExp(
                                                        IdnExp(IdnUse("x")),
                                                        IdnExp(IdnUse("y")))))),
                                        LetVarClause(
                                            Vector(IdnDef("h")),
                                            Vector(
                                                PlusExp(
                                                    MinusExp(
                                                        PlusExp(
                                                            NegExp(
                                                                SlashExp(
                                                                    StarExp(
                                                                        IntExp(10),
                                                                        IdnExp(IdnUse("a"))),
                                                                    IdnExp(IdnUse("b")))),
                                                            IdnExp(IdnUse("c"))),
                                                        NegExp(IntExp(10))),
                                                    StarExp(IdnExp(IdnUse("d")), IdnExp(IdnUse("e")))))),
                                        LetVarClause(
                                            Vector(IdnDef("k")),
                                            Vector(
                                                EqvExp(
                                                    XorExp(
                                                        FalseExp(),
                                                        OrExp(
                                                            OrExp(
                                                                AndExp(
                                                                    IdnExp(IdnUse("f")),
                                                                    IdnExp(IdnUse("g"))),
                                                                AndExp(
                                                                    IdnExp(IdnUse("a")),
                                                                    IdnExp(IdnUse("h")))),
                                                            IdnExp(IdnUse("c")))),
                                                    PlusExp(IdnExp(IdnUse("i")), IdnExp(IdnUse("j")))))),
                                        LetVarClause(
                                            Vector(IdnDef("l")),
                                            Vector(
                                                MinusExp(
                                                    PlusExp(
                                                        BinaryPlingExp(
                                                            IdnExp(IdnUse("v")),
                                                            IdnExp(IdnUse("i"))),
                                                        BinaryPlingExp(
                                                            IdnExp(IdnUse("v")),
                                                            IdnExp(IdnUse("j")))),
                                                    StarExp(IdnExp(IdnUse("a")), IdnExp(IdnUse("b"))))))))),
                            Vector(
                                ResultIsStmt(
                                    IfExp(
                                        IdnExp(IdnUse("a")),
                                        IdnExp(IdnUse("b")),
                                        IfExp(
                                            IdnExp(IdnUse("c")),
                                            IdnExp(IdnUse("d")),
                                            IdnExp(IdnUse("e")))))))))))))
